codegen/templates/dotnet.dotnet.erb

```diff
@@ -71,83 +67,14 @@ public sealed class <%= class_name(key) %>
     {
       <%- schema['properties'].each do |property_name, property|
           required = (schema['required'] || []).index(property_name) 
+          isValueType = type_for(class_name(key), property_name, property) == 'long' || type_for(class_name(key), property_name, property) == 'bool' || property['enum'] != nil
       -%>
-      <%- if required -%>
-        RequireNonNull<<%= type_for(class_name(key), property_name, property) %>>(<%= property_name %>, "<%= capitalize(property_name) %>", "<%= class_name(key) %>.<%= capitalize(property_name) %> cannot be null");
-      <%- end -%>
-      <%- if property['items'] -%>
-        <%- if required -%>
-        this.<%= capitalize(property_name) %> = new <%= type_for(class_name(key), property_name, property) %>(<%= property_name %>);        
-        <%- else -%>
-        this.<%= capitalize(property_name) %> = <%= property_name %> == null ? null : new <%= type_for(class_name(key), property_name, property) %>(<%= property_name %>);
-        <%- end -%>
+      <%- if !isValueType && required -%>
+        <%= capitalize(property_name) %> = <%= property_name %> ?? throw new ArgumentNullException("<%= capitalize(property_name) %>", "<%= class_name(key) %>.<%= capitalize(property_name) %> cannot be null");
```

Comment:
The template was generating "record-like" classes with getters only; in a sense providing an 'immutable' data structure. This PR is adopting the use of records directly, making a stronger statement about immutability.
During the constructor, we previously copyied the contents of enumerable items by calling 'new List<T>(argument) for those properties that were enumerable. This disconnected the Message instance from its source (if one list or the other got changed, the other would not).

This proposed template simply assigns the input argument to the generated property.

@gasparnagy How careful do we need to be about immutability and/or effects of change here?
We could use an ImmutableList<T> here to be explicit about intent and strengthen the behavior, go back to previous behavior of simply copying the list content, or accept this behavior of accepting the List argument as-is.

Reply:
Immutable array is desirable here, but i will require 1 extra dependency for .NET Standard tfm

Reply:
I think regardless of what structure we use to store the "immutable" list, we should go back to the previous model of "copying" the items from the input structure, so the simple assignment is not enough. 

As this is package is not really used yet, maybe we should be more future proof and go for the ImmutableList accepting the fact that this way we have to add a dependency to `System.Collections.Immutable` for the .NET Standard 2.0 target (in .NET 8 it is anyway included). So I would say, let's give it a try. If we see problems with this dependendency, we can still roll back.

Reply:
PR is updated

Reply:
Thx. As far as I see now it takes `ImmutableArray` as an input, so it will be the responsibility of the caller to ensure the immutability. I think this is good, but I wonder if we should generate an additional convenience overloads for the ctors that just take an `IEnumerable<T>` and build the `ImmutableArray` from it.
@clrudolphi Would creating an `ImmutableArray` for calling the ctors cause any issues for you?

Reply:
I would prefer the constructor signature to accept the IEnumerable<T> as a parameter, and have it create the ImutableArray property.

Reply:
I disagree that we need to make a copy inside of constructor without any visible reason. 
Maybe you have some code examples and I can propose a better solution?

Reply:
okay, then lets continue to use current solution with List copy